### PR TITLE
Force field parsing behavior

### PR DESCRIPTION
**Pull request by jan-stevens**
Only parse files with known extensions in force field directory.

### CODE REVIEW

File: polyply/src/load_library.py [hunk 1]
```diff
@@ -71,12 +71,9 @@ def wrapper(parser, path, force_field):
         path = pathlib.Path(path)
         file_extension = path.suffix.casefold()[1:]
 
-        try:
+        if file_extension in FORCE_FIELD_PARSERS.keys():
             parser = FORCE_FIELD_PARSERS[file_extension]
             wrapper(parser, path, force_field)
```
Comment (fgrunewald) [inline polyply/src/load_library.py:76]: I think we might want to raise a LOGGER.warning when the file extension is not know.

File: polyply/tests/test_load_library.py [hunk 2]
```diff
@@ -22,13 +22,17 @@
...
+    assert force_field.blocks
+    assert force_field.links
 
-#class TestLoadLibrary():
-#
-#      @staticmethod
-#      def test__resolve_lib_paths():
-#          libnames = ["martini", "2016H66"]
-#          paths = 
-#          _resolve_lib_paths(lib_names, DATA_PATH):
```
Comment (fgrunewald) [inline polyply/tests/test_load_library.py:34]: Should this be commented out?

File: polyply/tests/test_load_library.py [hunk 3]
```diff
@@ -22,13 +22,17 @@
...
-#from polyply import DATA_PATH
+def test_read_ff_from_files():
+    name = "ff"
+    force_field = vermouth.forcefield.ForceField(name)
+    lib_files = _resolve_lib_paths([name], TEST_DATA)
+    read_ff_from_files(lib_files, force_field)
+
+    # Check if .ff files were parsed
+    assert force_field.blocks
+    assert force_field.links
```
Comment (fgrunewald) [inline polyply/tests/test_load_library.py:37]: I think the test now parses the bld file but doesn't actually read it. my point being you'd need to test that the templates are actually there.

**Review by fgrunewald — CHANGES_REQUESTED** [review]

File: polyply/src/load_library.py [hunk 4]
```diff
@@ -101,12 +105,64 @@ def load_library(name, lib_names, extra_files):
...
+    topology: :class:`polyply.src.topology`
+
+    Returns
+    -------
+
+    """
+    for path in paths:
+        with open(path, 'r') as file_:
+            lines = file_.readlines()
+            read_build_file(lines, topology.molecules, topology)
```
Comment (fgrunewald) [inline polyply/src/load_library.py:139]: indent so that the read_file is in the same column as the open statement. We don't need the file to be open during processing of the lines.

**Review by fgrunewald — CHANGES_REQUESTED** [review]
Overall you did a good job on this. There were some smaller non-answered comments from earlier. If they don't apply you can resolve them, I don't remember. Otherwise I've 2 points:

1) I would prefer to rewrite the resolve lib paths in a way that it uses pathlib and returns a pathlib.Path object. That is the more modern way to do it and also more robust on other operating systems. This is more a general maintenance thing. 
2) files of unkown extension are now just skipped. However, in gen_seq and gen_itp it would still be useful to raise an error when the input file provided by the user has an unknown extension. I propose to implement the function below and call it before loading the library in both gen_seq and gen_itp. 

```
def check_extensions_ff(paths):
     for path in paths:
          file_extension = path.suffix.casefold()[1:]
          if file_extension not in ForceFieldParsers:
                 "Cannot parse file with extension {}".format(extension)
                 raise IOErr(msg)
```

File: polyply/src/load_library.py [hunk 5]
```diff
@@ -17,14 +17,50 @@
...
+    """
+    check the extensions of the
+    user provided build files
+
+    Parameters
+    ----------
+    bld_files: list[`pathlib.PosixPath`]
+        list of build files
+    """
+    for _file in bld_files:
```
Comment (pckroon) [inline polyply/src/load_library.py:41]:
```suggestion
    for file_ in bld_files:
```
PEP8 says to post-fix underscores in these cases

File: polyply/src/load_library.py [hunk 6]
```diff
@@ -35,13 +71,17 @@ def _resolve_lib_paths(lib_names, data_path):
...
     files = []
     for name in lib_names:
         directory = os.path.join(data_path, name)
         for _file in os.listdir(directory):
-            _file = os.path.join(directory, _file)
-            files.append(_file)
+            _, file_extension = os.path.splitext(_file)
+            if file_extension[1:] in allowed_extensions:
+                _file = os.path.join(directory, _file)
+                files.append(_file)
```
Comment (pckroon) [inline polyply/src/load_library.py:84]: I'm not sure this is the right place for this. Either don't filter here, add all files to `files` irrespective of extension, and let the actual parser-selector blow up; *or* filter based on the keys in `FORCE_FIELD_PARSERS` (and either way, do some `str.casefold`)

File: polyply/src/load_library.py [hunk 7]
```diff
@@ -51,8 +91,8 @@ def read_ff_from_files(paths, force_field):
 
     Parameters
     ----------
-    paths: list
-           List of vaild file paths
+    paths: list[`os.path`]
+           List of valid file paths
```
Comment (pckroon) [inline polyply/src/load_library.py:95]:
```suggestion
    paths: list[`pathlib.Path`]
           List of valid file paths
```

File: polyply/src/load_library.py [hunk 8]
```diff
@@ -71,17 +111,13 @@ def wrapper(parser, path, force_field):
...
         file_extension = path.suffix.casefold()[1:]
 
-        try:
-            parser = FORCE_FIELD_PARSERS[file_extension]
-            wrapper(parser, path, force_field)
-        except KeyError:
-            raise IOError(
-                "Cannot parse file with extension {}.".format(file_extension))
+        parser = FORCE_FIELD_PARSERS[file_extension]
+        wrapper(parser, path, force_field)
```
Comment (pckroon) [inline polyply/src/load_library.py:115]: EAFP: Easier to ask for forgiveness  than for permission. Revert this change

File: polyply/src/load_library.py [hunk 9]
```diff
@@ -90,9 +126,9 @@ def load_library(name, lib_names, extra_files):
     ----------
     name: str
       Force-field name
-    lib_names: list
+    lib_names: list[`os.path`]
       List of lirbary names
-    extra_files: list
+    extra_files: list[`os.path`]
```
Comment (pckroon) [inline polyply/src/load_library.py:131]: Same as above, pathlib.Path. os.path is a module, not a path-like object

File: polyply/src/load_library.py [hunk 10]
```diff
@@ -101,12 +137,64 @@ def load_library(name, lib_names, extra_files):
...
-        lib_files = _resolve_lib_paths(lib_names, DATA_PATH)
+        lib_files = _resolve_lib_paths(lib_names, DATA_PATH,
+                                       FORCE_FIELD_PARSERS.keys())
         all_files = lib_files + extra_files
     elif lib_names:
-        all_files = _resolve_lib_paths(lib_names, DATA_PATH)
+        all_files = _resolve_lib_paths(lib_names, DATA_PATH,
+                                       FORCE_FIELD_PARSERS.keys())
     elif extra_files:
         all_files = extra_files
```
Comment (pckroon) [inline polyply/src/load_library.py:147]: `all_files = _resolve_lib_paths(lib_names, DATA_PATH) + extra_files`
replaces this entire tree :)
Assuming of course that `lib_names` and `extra_files` are something sane such as empty lists when they're False-ish. And assuming that _resolve_lib_paths returns a list.

File: polyply/src/load_library.py [hunk 11]
```diff
@@ -101,12 +137,64 @@ def load_library(name, lib_names, extra_files):
...
+    topology: :class:`polyply.src.topology`
+
+    Returns
+    -------
+
+    """
+    for path in paths:
+        with open(path, 'r') as file_:
+            lines = file_.readlines()
+        read_build_file(lines, topology.molecules, topology)
```
Comment (pckroon) [inline polyply/src/load_library.py:171]:
```suggestion
    for path in paths:
        with open(path, 'r') as file_:
            read_build_file(file_, topology.molecules, topology)
```
Don't read the entire file in memory if you can avoid it. `file_` is an iterator.

File: polyply/src/load_library.py [hunk 12]
```diff
@@ -101,12 +137,64 @@ def load_library(name, lib_names, extra_files):
...
+        lib_files = _resolve_lib_paths(lib_names, DATA_PATH, ['bld'])
+        all_files = lib_files + [build_file]
+    elif lib_names:
+        all_files = _resolve_lib_paths(lib_names, DATA_PATH, ['bld'])
+    elif build_file:
+        all_files = [build_file]
+    else:
+        return
+
+    read_build_options_from_files(all_files, topology)
```
Comment (pckroon) [inline polyply/src/load_library.py:200]:
```suggestion
    all_files = _resolve_lib_paths(lib_names, DATA_PATH, ['bld'])
    if build_file:
         all_files += [build_file]
    
    read_build_options_from_files(all_files, topology)
```
But do you need this if? In what scenario do I call this function without a build_file?

**Review by pckroon — COMMENTED** [review]
`check_extension_bld` and `check_extensions_ff` do the same thing and should be folded into one. 
How thoroughly you need to restructure will depend on the *intention* behind bld and ff files. As I understand it, bld adds to a Topology, and ff adds to a ForceField.
Every Topology has  a .forcefield attribute, so you could give all parsers a Topology object rather than splitting it between bld parsers (who need a Topology) and ff parsers (who need a ForceField). You'd have to make small wrappers for the current ff parsers that extract the ForceField from the Topology they'd get. In other words, give all parser functions the same "storage_object" as Fabian calls it (a Topology), and let the parsers deal with it.

Doesn't scale indefinitely, but good enough for now?
Finally, double check that all `parse_*`, `read_*` and `load_*` have the same signature! (Insofar you can't fold them in one) It'll save you a lot of headaches.

**Comment by jan-stevens** [comment]
I have reduced the complexity of the force field and build file-parsing wrappers. 

The wrappers can be further simplified by merging `load_ff_library` and `load_build_files`, this would however require the `load_ff_library` to take a forcefield object as input and not a simply a name. (i.e. in `gen_itp.py`, an empty forcefield has to be created and provided to `load_ff_library`. ) I guess this would also be okay, opinion @fgrunewald?

p.s I will fix the relevant tests later if the layout of the wrappers is okay.

File: polyply/src/build_file_parser.py [hunk 13]
```diff
@@ -260,7 +260,7 @@ def _base_parser_geometry(tokens, _type):
         geometry_def["parameters"] = parameters
         return geometry_def
 
-def read_build_file(lines, molecules, topology):
+def read_build_file(lines, topology, molecules=None):
```
Comment (fgrunewald) [inline polyply/src/build_file_parser.py:263]: I think we could even git rid of molecules there. Even though probably a lot of test fail then

File: polyply/src/build_file_parser.py [hunk 14]
```diff
@@ -271,5 +271,8 @@ def read_build_file(lines, molecules, topology):
         list of lines of an itp file
     force_field: :class:`vermouth.forcefield.ForceField`
```
Comment (fgrunewald) [inline polyply/src/build_file_parser.py:272]: Doctoring needs updating.

File: polyply/src/load_library.py [hunk 15]
```diff
@@ -13,75 +13,109 @@
...
+    file_: class:`pathlib.PosixPath`
+        path to file
+    file_parsers: dict
+        dictionary of available file parsers
+    """
+    file_extension = file_.suffix[1:]
+    if file_extension not in file_parsers:
+        msg = "Cannot parse file file with extension {}".format(file_extension)
+        raise IOError(msg)
+    else:
```
Comment (fgrunewald) [inline polyply/src/load_library.py:46]: no need for the else here; just return

File: polyply/src/load_library.py [hunk 16]
```diff
@@ -13,75 +13,109 @@
...
 from polyply import DATA_PATH
 from .ff_parser_sub import read_ff
+from .build_file_parser import read_build_file
 from .polyply_parser import read_polyply
 
+LOGGER = StyleAdapter(get_logger(__name__))
+BUILD_FILE_PARSERS = {'bld': read_build_file}
 FORCE_FIELD_PARSERS = {'rtp': read_rtp, 'ff': read_ff, 'itp': read_polyply, 'bib': read_bib}
 
+def determine_parser(file_, file_parsers):
```
Comment (fgrunewald) [inline polyply/src/load_library.py:30]: let's call this thing `get_parser` or something along those lines to keep it a bit shorter

File: polyply/src/load_library.py [hunk 17]
```diff
@@ -13,75 +13,109 @@
...
-           List of vaild file paths
-    force_field: class:`vermouth.force_field.ForceField`
-
-    Returns
-    -------
-    force_field: class:`vermouth.force_field.ForceField`
-       updated forcefield
+    paths: list[`pathlib.Path`]
+           List of provided file paths
+    storage_object: topology or forcefield
```
Comment (fgrunewald) [inline polyply/src/load_library.py:80]: add reference to the full object like:

`vermouth.ForceField` and `polypyl.src.topology.Topology` 

If we ever make a documentation Sphinx will work better on those

File: polyply/src/load_library.py [hunk 18]
```diff
@@ -13,75 +13,109 @@
...
+    """
+    Load build file options and molecule templates into topology.
+
+    Parameters
+    ----------
+    topology: :class:`polyply.src.topology`
+    build_file: str
+        List of build files to parse
+    lib_names: list[str]
+        List of library names for which to load templates
```
Comment (fgrunewald) [inline polyply/src/load_library.py:107]: align order with order in function args

File: polyply/src/load_library.py [hunk 19]
```diff
@@ -13,75 +13,109 @@
...
+    ----------
+    file_: class:`pathlib.PosixPath`
+        path to file
+    file_parsers: dict
+        dictionary of available file parsers
+    """
+    file_extension = file_.suffix[1:]
+    if file_extension not in file_parsers:
+        msg = "Cannot parse file file with extension {}".format(file_extension)
+        raise IOError(msg)
```
Comment (fgrunewald) [inline polyply/src/load_library.py:45]: Is it worth making this a custom but more clear exception? @pckroon ideas

File: polyply/src/gen_coords.py [hunk 20]
```diff
@@ -115,6 +115,7 @@ def gen_coords(toppath,
                step_fudge=1.0,
                max_force=5*10**4.0,
                nrewind=5,
+               lib=None,
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/gen_coords.py:118]: I don't dislike this but also not a fan still ... We'll see

**Review by fgrunewald — CHANGES_REQUESTED** [review]
I've added some smaller nitpick comments. Also @jan-stevens please resolve the comment that don't apply anymore from the previous round of review. 

Overall I think this PR is very cool. It makes the loading of files more streamline, and changes the path handling to all `pathlib` based functions.
 
Regarding the additional combining of loaders. For now I think we should keep the load_forcefield and load_build file separated. The topology object only occurs in workflows related to coordinate building and needs itself some trimming and flashing out. Thus I would prefer to keep it out of `gen_params` at the moment. We should make an issue of it though, to keep track. 

There is one more thing, which we need to test. There is a hierarchy of files. That is CLI provided files should overwrite definitions provided in the library. From my looking at the code, I'd say this is fine. However, it would be really important to add that test.

File: polyply/src/load_library.py (hunk 10 above)
Comment (jan-stevens) [reply polyply/src/load_library.py]: Nice, that removed a lot of clutter!

File: polyply/src/gen_coords.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jan-stevens) [reply polyply/src/gen_coords.py:118]: By default the force_field as the name of the molecule that is generated. As far as I know it is not possible to get the name of the force field from the actual force_field attribute of the Topolgy, right?

Because in that case we can just read it from the force field.

File: polyply/src/build_file_parser.py (hunk 13 above)
Comment (jan-stevens) [reply polyply/src/build_file_parser.py:263]: Couldn't there in the future be a case where you want to only apply certain bld file setting to specific molecules in a topology? (i.e. topologies of DNA + something else)

**Comment by jan-stevens** [comment]
I realized that when loading the buildfiles from the force field directory, the get_parser() function would complain about the force field files that it encounters.

Does the fix in the `FORCE_FIELD_PARSERS' make sense? @fgrunewald

**Comment by jan-stevens** [comment]
it was a difficult delivery, but I think the library loading is now fixed. Can you have a look at it @fgrunewald ?

File: polyply/src/load_library.py [hunk 21]
```diff
@@ -13,75 +13,112 @@
...
+        dictionary of available file parsers
+    is_lib_file: bool
+        indicates whether the provided path is from a data library
+    """
+    file_extension = file_.suffix[1:]
+    if file_extension in file_parsers:
+        return file_parsers[file_extension]
+    elif not is_lib_file:
+        msg = "Cannot parse user provided file with extension {file_extension}."
+        raise IOError(msg)
```
Comment (fgrunewald) [inline polyply/src/load_library.py:49]: It might be useful to add a third exception in case an unkown file is added to the library

File: polyply/src/load_library.py [hunk 22]
```diff
@@ -13,75 +13,112 @@
...
     ----------
-    paths: list
-           List of vaild file paths
-    force_field: class:`vermouth.force_field.ForceField`
-
-    Returns
-    -------
-    force_field: class:`vermouth.force_field.ForceField`
-       updated forcefield
+    paths: list[`pathlib.Path`]
```
Comment (fgrunewald) [inline polyply/src/load_library.py:80]: based on line 93 paths needs to contain exactly 2 lists with lib_files and user_files. We should add some more information to the docstring about the requirements

File: polyply/src/load_library.py [hunk 23]
```diff
@@ -13,75 +13,112 @@
...
 
-    for path in paths:
-        path = pathlib.Path(path)
-        file_extension = path.suffix.casefold()[1:]
+    lib_files, user_files = paths
+    for path in user_files + lib_files or []:
+        is_lib_file = path in lib_files
+        parser = get_parser(path, file_parsers, is_lib_file)
+        if parser:
+            parse_file(parser, path, storage_object)
```
Comment (fgrunewald) [inline polyply/src/load_library.py:98]: just for consistency this function should probably return the storage_object

File: polyply/src/load_library.py [hunk 24]
```diff
@@ -13,75 +13,112 @@
...
 
-    return force_field
+def load_build_files(topology, lib_names, build_file):
+    """
+    Load build file options and molecule templates into topology.
+
+    Parameters
+    ----------
+    topology: :class:`polyply.src.topology`
+    lib_names: list[str]
```
Comment (fgrunewald) [inline polyply/src/load_library.py:108]: should this be able to handle mutliple ffs. What if they contain conflicting information?

File: polyply/src/load_library.py [hunk 25]
```diff
@@ -13,75 +13,112 @@
...
+    Returns
+    -------
+
+    """
+    all_files = [_resolve_lib_files(lib_names, DATA_PATH), build_file]
+    read_options_from_files(all_files, topology, BUILD_FILE_PARSERS)
 
 
-def load_library(name, lib_names, extra_files):
+def load_ff_library(name, lib_names, extra_ff_file):
```
Comment (fgrunewald) [inline polyply/src/load_library.py:121]: I think you should add the force-field as argument as well. Then the function either creates a new force-field or populates an existing one. Both options are handy I think

File: polyply/src/load_library.py (hunk 19 above)
Comment (pckroon) [reply polyply/src/load_library.py]: Stick to the built-in exceptions. If needed clarify the message

File: bin/polyply [hunk 26]
```diff
@@ -97,8 +97,10 @@ def main(): # pylint: disable=too-many-locals,too-many-statements
                             help='input file molecules (.gro)', default=None)
     file_group.add_argument('-mc', dest='coordpath_meta', type=Path,
                             help='input file meta-molecule (.gro)', default=None)
-    file_group.add_argument('-b', dest='build', type=Path, default=None,
+    file_group.add_argument('-b', dest='build', type=Path, required=False, default=[], nargs='*',
                             help=('input file; specify molecule specific building options'))
+    file_group.add_argument('-lib', dest='lib', required=False, type=Path, default=[],
+                            help='force-fields to include from library', nargs='*')
```
[the remark this answers is not in the document]
Comment (jan-stevens) [reply bin/polyply:103]: Currently, if the force field does not contain any templates nothing happens. I also think it would be good to add a warning here.

**Review by fgrunewald — COMMENTED** [review]
looks good to me overall. What was the issue with the test? 

Can you run the regression tests on this once more?

**Comment by jan-stevens** [comment]
I copied a code snippet that captures and checks if the logger threw the correct Warning from some other test in the codebase. But for some reason, this caused `pkg_resources` to throw a DeprecationWarning only when the tests were run by GH and this broke the test somehow. I could not reproduce the error locally, but a nicer way of capturing the Warning fixed the issue.

And the regression tests all passed btw.

**Review by fgrunewald — APPROVED** [review]

File: polyply/src/gen_coords.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/gen_coords.py:118]: this would be the best solution but I think there is no universal way to do this

[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/gen_coords.py:118]: @jan-stevens please open this as an issue for improvement
